### PR TITLE
fix: refine shouldUseLegacyApi and add tests

### PR DESCRIPTION
**Pull request by villebro**
### SUMMARY
The bug that #10147 fixes exposed a shortcoming in the `shouldUseLegacyApi` function in `exploreUtils`. This fixes the bug and adds unit tests to catch all possible cases I could think of.

### TEST PLAN
CI + new tests

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/spec/javascripts/explore/utils_spec.jsx [hunk 1]
```diff
@@ -202,4 +204,45 @@ describe('exploreUtils', () => {
...
+      getChartMetadataRegistry()
+        .registerValue('my_legacy_viz', { useLegacyApi: true })
+        .registerValue('my_v1_viz', { useLegacyApi: false });
+    });
+
+    afterAll(() => {
+      getChartMetadataRegistry().remove('my_legacy_viz').remove('my_v1_viz');
+    });
+
+    it('return true for legacy viz', () => {
```
Comment (rusackas) [inline superset-frontend/spec/javascripts/explore/utils_spec.jsx:219]: NIT, but maybe "returns" for these so it('reads like a sentence')

**Comment by villebro** [comment]
@rusackas thanks for nittery, comments addressed!

**Review by rusackas — APPROVED** [review]
image

**Review by bkyryliuk — APPROVED** [review]
LGTM !

**Pull request merged**